### ISSUE TITLE
What can i do if i want to use focalloss in SSD model?

### ISSUE THREAD

**Issue by zqburde** [dmlc/gluon-cv#220]
I have see focalloss that has been written in the loss.py, now  i want to use it,  is it possible for me to call it directly or modify SSDloss？

**Comment by zhreshold** [comment]
Replace the cross entropy loss in MultiBoxLoss

**Comment by zqburde** [comment]
@zhreshold Thanks

**Comment by Angzz** [comment]
@zqburde Hi, have you successfully use focal loss in gluoncv?

**Comment by Angzz** [comment]
@zhreshold I see ''compute element-wise cross entropy loss and sort, then perform negative mining'' in the MultiboxLoss description, is it equal to say if I use focal loss, the sort and nagative mining is not needed?

**Comment by Angzz** [comment]
This is my settings in loss.py, but the loss is always nan.

focal_loss = FocalLoss(axis=-1, from_logits=True, num_class=81, size_average=True)
        for cp, bp, ct, bt in zip(*[cls_pred, box_pred, cls_target, box_target]):
            pred = nd.log_softmax(cp, axis=-1)
            pos = ct > 0

            '''
            cls_loss = -nd.pick(pred, ct, axis=-1, keepdims=False)
            rank = (cls_loss * (pos - 1)).argsort(axis=1).argsort(axis=1)
            hard_negative = rank < (pos.sum(axis=1) * self._negative_mining_ratio).expand_dims(-1)
            # mask out if not positive or negative
            cls_loss = nd.where((pos + hard_negative) > 0, cls_loss, nd.zeros_like(cls_loss))
            '''

            cls_loss = focal_loss(pred, ct)

            # cls_losses.append(nd.sum(cls_loss, axis=0, exclude=True) / num_pos_all)
            cls_losses.append(cls_loss)

            bp = _reshape_like(nd, bp, bt)
            box_loss = nd.abs(bp - bt)
            box_loss = nd.where(box_loss > self._rho, box_loss - 0.5 * self._rho,
                                (0.5 / self._rho) * nd.square(box_loss))
            # box loss only apply to positive samples
            box_loss = box_loss * pos.expand_dims(axis=-1)
            box_losses.append(nd.sum(box_loss, axis=0, exclude=True) / num_pos_all)
            sum_losses.append(cls_losses[-1] + self._lambd * box_losses[-1])

        return sum_losses, cls_losses, box_losses

**Comment by zhreshold** [comment]
@Angzz Since you are using ndarray, I would suggest print intermediate variables one by one to evaluate the correctness of these values.

**Comment by Angzz** [comment]
@zhreshold Thanks very much, it works for me with the following code, but I want to please you have a look whether it is reasonable, I just show part of the code below, thanks.

        focal_loss = FocalLoss(num_class=21)
        for cp, bp, ct, bt in zip(*[cls_pred, box_pred, cls_target, box_target]):
            '''Focal Loss'''
            pred = nd.log_softmax(cp, axis=-1)
            pos = ct > 0
            cls_loss = focal_loss(pred, ct) # [16, 31467, 81]
            cls_loss = nd.sum(cls_loss, axis=-1) # [16, 31467]
            rank = (cls_loss * (pos - 1)).argsort(axis=1).argsort(axis=1)
            hard_negative = rank < (pos.sum(axis=1) * self._negative_mining_ratio).expand_dims(-1)
            # mask out if not positive or negative
            cls_loss = nd.where((pos + hard_negative) > 0, cls_loss, nd.zeros_like(cls_loss))
            cls_losses.append(nd.sum(cls_loss, axis=0, exclude=True) / num_pos_all)


            bp = _reshape_like(nd, bp, bt)
            box_loss = nd.abs(bp - bt)
            box_loss = nd.where(box_loss > self._rho, box_loss - 0.5 * self._rho,
                                (0.5 / self._rho) * nd.square(box_loss))
            # box loss only apply to positive samples
            box_loss = box_loss * pos.expand_dims(axis=-1)
            box_losses.append(nd.sum(box_loss, axis=0, exclude=True) / num_pos_all)
            sum_losses.append(cls_losses[-1] + self._lambd * box_losses[-1])

        return sum_losses, cls_losses, box_losses

With train_set = voc07trainval + voc12trainval, test_set = voc07test, the focal_loss starts with 1.9XXX compared with CrossEntropy 8.4XXX, I will show you the performence each other later.

**Comment by Angzz** [comment]
I forget to mention I return the loss directly without ``self._size_average`` settings in class ``FocalLoss``

    def hybrid_forward(self, F, pred, label, sample_weight=None):
        """Loss forward"""
        if not self._from_logits:
            pred = F.sigmoid(pred)
        if self._sparse_label:
            one_hot = F.one_hot(label, self._num_class)
        else:
            one_hot = label > 0
        pt = F.where(one_hot, pred, 1 - pred)
        t = F.ones_like(one_hot)
        alpha = F.where(one_hot, self._alpha * t, (1 - self._alpha) * t)
        loss = -alpha * ((1 - pt) ** self._gamma) * F.log(F.minimum(pt + self._eps, 1))
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        
        return loss
        #if self._size_average:
        #    return F.mean(loss, axis=self._batch_axis, exclude=True)
        #else:
        #    return F.sum(loss, axis=self._batch_axis, exclude=True)

**Comment by zhreshold** [comment]
Just a quick examine, you should not use nd.log_softmax together with F.sigmoid() in FocalLoss, the rest looks good

**Comment by Angzz** [comment]
@zhreshold Thanks very much, but it seems that the performence is not as good as CrossEntropy...

**Comment by TriLoo** [comment]
@Angzz 
Could you plz give me some explanations about the implementation of HNM, `i.e.`

```Python
        rank = (cls_loss * (pos - 1)).argsort(axis=1).argsort(axis=1)
        hard_negative = rank < (pos.sum(axis=1) * self._negative_mining_ratio).expand_dims(-1)
        # mask out if not positive or negative
        cls_loss = nd.where((pos + hard_negative) > 0, cls_loss, nd.zeros_like(cls_loss))
        cls_losses.append(nd.sum(cls_loss, axis=0, exclude=True) / num_pos_all)

After this process, I found that most elements in `cls_closs`  were set to zero. This can be the reason why the focal loss is smaller than cross entropy in  your case.
